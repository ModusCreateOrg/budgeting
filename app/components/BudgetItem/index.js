// @flow
import * as React from 'react';
import formatAmount from 'utils/formatAmount';

import { Route, Link } from 'react-router-dom';
import DonutChart from 'components/DonutChart';
import styles from './style.scss';

type BugdetDetailsProps = {
  transaction: Transaction,
  contribution: Object,
};

const BudgetItem = ({ transaction, contribution }: BugdetDetailsProps) => {
  const chartData = [
    {
      value: parseFloat(contribution.percent),
      label: 'This',
    },
    {
      value: 100 - parseFloat(contribution.percent),
      label: 'Total',
    },
  ];
  return (
    <section>
      <Route>
        <Link to="/budget">Back</Link>
      </Route>
      <h1>
        {transaction.description}
        &nbsp;({formatAmount(transaction.value).text})
      </h1>
      <h3>
        <span className={transaction.value < 0 ? styles.redText : styles.greenText}>
          {transaction.value > 0 ? '+' : '-'}
        </span>
        {formatAmount(contribution.percent, false, true).text}
      </h3>
      <DonutChart data={chartData} dataLabel="label" dataKey="label" valueAsPercent="true" />
    </section>
  );
};

export default BudgetItem;
