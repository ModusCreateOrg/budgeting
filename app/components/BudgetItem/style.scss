@import 'theme/main';

h1,
h4 {
	text-align: center;
}
.neg {
	color: $red;
}
.pos {
	color: $green;
}

.navLink {
	text-decoration: none;
	font-size: 1.2em;
	color: $white;
	padding: 10px;
	display: inline-block;
	background-color: $darkgreen;
	border-radius: 10px;

	&:link,
	&:visited {
		color: $white;
	}
}

.buttonWrapper {
	text-align: center;
	margin: 30px auto;
	padding: 10px;
}
