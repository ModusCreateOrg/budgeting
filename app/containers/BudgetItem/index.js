import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { getCategories } from 'selectors/categories';
import {
    getTransactions,
    getInflowBalance,
    getOutflowBalance
  } from 'selectors/transactions';
import formatAmount from 'utils/formatAmount';
import type { Transaction } from 'modules/transactions';
import type { Categories } from 'modules/categories';
import PieChart from 'components/PieChart';
import styles from './style.scss';
import backbutton from 'images/back.svg';

class BudgetItem extends React.Component {
    render() {
        let itemId = history.state.state.id;
        let transaction = this.props.transactions.filter(function(item){
            return item.id == itemId
        })[0];    
        const amount = formatAmount(transaction.value);
        let amountCls, chartFillColor;
        const sign = amount.isNegative ? '-' : '+';
        let percentage = 0;
        if(amount.isNegative){
            percentage = ((parseFloat(transaction.value)/this.props.outflow) * 100).toFixed(2);
            amountCls = styles.neg;
            chartFillColor = "#eb2a2a";
        }
        else{
            amountCls = styles.pos;
            percentage = ((parseFloat(transaction.value)/this.props.inflow) * 100).toFixed(2);
            chartFillColor = "#189c2d";
        }
        const category = this.props.categories[transaction.categoryId];
        let chartData = [];
        let otherColor = "#000";
        chartData.push({title: "Other", value: 100-(+(percentage)), color: otherColor})
        chartData.push({title: category, value: +(percentage), color: chartFillColor});
        return (
            <div className={styles.budgetItem}>
                <Link to="/budget" className={styles.back}><img src={backbutton} /></Link>
                <h1>{category + ' - ' + transaction.description + ' (' + amount.text+ ')'}</h1>
                <h2 className={amountCls}><span>{sign}</span><span>{percentage+"%"}</span></h2>
                <PieChart data={chartData} />;
            </div>
        );
    }
}

const mapStateToProps = state => ({
    transactions : getTransactions(state),
    categories : getCategories(state),    
    inflow : getInflowBalance(state),
    outflow : getOutflowBalance(state)
  });
export default connect(mapStateToProps)(BudgetItem);