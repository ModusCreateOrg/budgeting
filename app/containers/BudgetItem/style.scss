@import 'theme/variables';
.budgetItem {
  display: flex;
  flex-flow: column;
  flex: 1;
  border: none;
  font-size: 0.8em;
  width: 80%;
  h1,
  h3 {
    margin: 5px;
  }
  .neg {
    color: $red;
  }
  .pos {
    color: $green;
  }
  .back {
    text-decoration: none;
    background-color: #f1f1f1;
    border-radius: 50%;
    margin: 5px;
    width: 5%;
    img{
      width: 100%;
      height: 100%;
    }
  }
  .back:hover {
    background-color: #ddd;
  }
  @media only screen and (max-width: $screen-small) {
    .back {
      width: 10%;
    }
  }
}
