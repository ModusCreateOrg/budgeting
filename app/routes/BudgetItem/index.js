// @flow
import React from 'react';
import Chunk from 'components/Chunk';

const loadBudgetItemContainer = () => import('components/BudgetItem' /* webpackChunkName: "budgetitem" */);

const BudgetItem = ({ match }) => <Chunk load={loadBudgetItemContainer} budgetItemId={match.params.id} />;

export default BudgetItem;
