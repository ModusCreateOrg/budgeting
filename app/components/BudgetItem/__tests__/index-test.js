import React from 'react';
import renderer from 'react-test-renderer';
import BudgetItem from 'components/BudgetItem';

it('renders correctly', () => {
  const mockTransaction = {
    id: 1,
    description: "Trader Joe's food",
    value: -423.34,
    categoryId: 1,
  };

  const mockContribution = {
    flowTotal: 1000,
    percent: 20,
  };

  const mockClick = () => {};

  const tree = renderer
    .create(<BudgetItem transaction={mockTransaction} contribution={mockContribution} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
