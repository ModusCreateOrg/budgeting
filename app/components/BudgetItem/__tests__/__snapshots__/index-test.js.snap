// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section>
  <h1
    className="h1"
  >
    Trader Joe's food
     (
    Groceries
    )
  </h1>
  <h4
    className="neg"
  >
    -$423.34
     (
    --9.41%
    )
  </h4>
  <div
    className="buttonWrapper"
  />
</section>
`;
